File: Haskell/Problem_026.hs
-- Find the value of d < 1000 for which 1/d contains the longest recurring cycle in its decimal fraction part.


inverseWithMaxRecurringUnder :: Integer -> Integer
inverseWithMaxRecurringUnder n =
  snd $ maximum $ zip [ lengthRecurring 1 x | x <- [2 .. n] ] [2 .. n]


lengthRecurring :: Integer -> Integer -> Int
lengthRecurring 0   _   = 0
lengthRecurring num den = _lengthRecurring num den []
 where
  _lengthRecurring num den rs = __lengthRecurring num den rs
   where
    rem = snd $ _divMod num den
    __lengthRecurring num den rs
      | rem == 0      = 0
      | rem `elem` rs = length rs
      | otherwise     = _lengthRecurring rem den (rem : rs)


_divMod :: Integer -> Integer -> (Integer, Integer)
_divMod n d = divModHelper n d
 where
  (q, r) = divMod n d
  divModHelper n d | q == 0    = _divMod (n * 10) d
                   | otherwise = (q, r)


main :: IO ()
main = do
  print $ inverseWithMaxRecurringUnder 1000

File: Python 3/Problem_026.py
# Find the value of d < 1000 for which 1/d contains the longest recurring cycle in its decimal fraction part.


def long_div(a, b):
    assert a >= 0 and b > 0
    if a == 0:
        return 0

    remainders = set()
    while True:

        quot, rem = divmod(a, b)
        while quot == 0:
            a *= 10
            quot, rem = divmod(a, b)

        if rem == 0:  # division is proper
            return 0

        elif rem in remainders:  # we reached a loop
            return len(remainders)

        else:  # we continue searching
            remainders.add(rem)
            a = rem


def longest_recurring_under(N):
    longest_ix = -1
    longest_length = 0
    for n in range(1, N):
        this_length = long_div(1, n)
        if this_length > longest_length:
            longest_ix, longest_length = n, this_length
    return longest_ix


if __name__ == "__main__":
    print(longest_recurring_under(1000))
